### PR TITLE
Core: state machine entity base class

### LINKED ISSUE #1463: Core: base classes for entities and in-memory store
# Feature Request

Developing an extension that introduces a custom entity with `state-machine-lib` requires writing boilerplate code that is already duplicated between `TransferProcess` and `ContractNegotiation`, as well as their store implementations (presumably this will apply to `DataFlowRequest` as well, when persistence for it is developed). There is an opportunity to introduce extensibility and reduce technical debt instead.

The code in question is primarily associated with state code and associated timestamps.

To reduce code duplication, we could develop:
- Generic base class for entities with associated state.
- Generic base class for in-memory entity store.
- Adapt `TransferProcess` and `ContractNegotiation` and their in-memory stores to use the base classes.

As follow-up work, generic base classes for Cosmos DB and PostgreSQL entity stores could be developed.

The interfaces and behavior would not change.

Note: `InMemoryTransferProcessStore` uses an additional caching layer and custom locking mechanism. Given this class is only meant to use for testing, it seems reasonable to remove this additional layer and use a simple `ConcurrentHashMap` for storage in the new generic in-memory store to reduce complexity.

## Which Areas Would Be Affected?

Core

## Solution Proposal

```java

public abstract class StateMachineEntity<T> implements TraceCarrier {

    protected String id;
    protected long createdTimestamp;
    protected int state;
    protected int stateCount;
    protected long stateTimestamp;
    protected Map<String, String> traceContext = new HashMap<>();
    protected String errorDetail;
    protected Clock clock;

   protected abstract static class Builder<T extends StateMachineEntity, B extends Builder<T, B>> {
   /...
}

public class TransferProcess extends StateMachineEntity<TransferProcess> {
}
```

```java
public class InMemoryEntityStore<T extends StateMachineEntity<T>> implements StateEntityStore<T> {
    T find(String id);
    void create(T entity);
    void update(T entity);
    void delete(String processId);
    Stream<T> findAll(QuerySpec querySpec);
    List<T> nextForState(int state, int max);
}
```

## Type of Issue

Technical debt refactoring

## Checklist

- [ ] assigned appropriate label?
- [x] **Do NOT select a milestone or an assignee!**

### PR DESCRIPTION
**Pull request by algattik**
## What this PR changes/adds

- Extracted `StateMachine` utility class for state machine entities
- Renamed previous `StateMachine` class to `StateMachineManager` for clarity and uniformity
- Developer documentation on how to use/develop persistent state machine
- Removed caching layer in in-memory store (as it added complexity with unclear benefits for code meant to be used for testing/at small scale)

## Why it does that

- Use the concept of [state machine](https://en.wikipedia.org/wiki/Finite-state_machine) in a way that follows its definition
- Reduce duplication in TransferProcess and ContractNegotiation and their in-memory stores
- Make state machine easier to use together with state-machine-lib in extension projects.

## Further notes

Also fixed race condition in ContractNegotiation (`transition(DECLINED, DECLINING, PROVIDER_OFFERED, CONFIRMING, CONFIRMED, REQUESTED);`), that manifested itself during a test run

## Linked Issue(s)

Closes #1463 

## Checklist

- [ ] added appropriate tests?
- [ ] performed checkstyle check locally?
- [ ] added/updated copyright headers?
- [ ] documented public classes/methods?
- [ ] added/updated relevant documentation?
- [ ] added relevant details to the changelog? (_skip with label `no-changelog`_)
- [ ] formatted title correctly? (_take a look at the [CONTRIBUTING](https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/blob/main/CONTRIBUTING.md#submit-a-pull-request) and [styleguide](https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/blob/main/styleguide.md) for details_)

### CODE REVIEW

**Review by jimmarino — APPROVED** [review]

**Review by ndr-brt — COMMENTED** [review]
What's the motivation behind modeling `StatefulEntity` as an interface and `StateMachine` as a superclass of `TransferProcess` and `ContractNegotiation`?
In my opinion they are not a `StateMachine`, they are `StatefulEntity` which state transitions are handled through a `StateMachine`, so the latter shouldn't be a superclass but something that interacts with the entities into the manager.

I would keep the `StatefulEntity` as an abstract class that would end up having an `Entity` superclass in turn being superclass of `TransferProcess` and `ContractNegotiation`.
The `Entity` class in the future would become the superclass of other non-stateful entities as `Asset` and `PolicyDefinition`

**Comment by algattik** [comment]
> What's the motivation behind modeling `StatefulEntity` as an interface and `StateMachine` as a superclass of `TransferProcess` and `ContractNegotiation`? In my opinion they are not a `StateMachine`, they are `StatefulEntity` which state transitions are handled through a `StateMachine`, so the latter shouldn't be a superclass but something that interacts with the entities into the manager.

Formally, "A state machine is an abstract machine that can be in exactly one of a finite number of states at any given time" (https://en.wikipedia.org/wiki/Finite-state_machine) so the state machine is the entity, not the manager.

Nevertheless, I have renamed `StateMachine` to `StatefulEntity`

> I would keep the `StatefulEntity` as an abstract class that would end up having an `Entity` superclass in turn being superclass of `TransferProcess` and `ContractNegotiation`. The `Entity` class in the future would become the superclass of other non-stateful entities as `Asset` and `PolicyDefinition`

As I understand this will be done by @diegogomez-zf as part of https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/issues/1370. Is it ok to do things in two steps like this?

**Review by ndr-brt — APPROVED** [review]

File: docs/developer/state-machine.md [hunk 1]
```diff
@@ -0,0 +1,68 @@
+# State machine
+
+EDC provides a lightweight framework to facilitate the development of persistent state machines.
+
+The framework currently manage a single execution thread.
```
Comment (paullatzelsperger) [inline docs/developer/state-machine.md:5]:
```suggestion
The framework currently manages a single execution thread.
```

File: docs/developer/state-machine.md [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+      return true;
+    }
+   return false;
+}
+...
+```
+
+## State-machine store
+
+The Store which manage `StatefulEntity` persistence must persist entities in a storage system. In-memory implementations are provided for testing.
```
Comment (paullatzelsperger) [inline docs/developer/state-machine.md:55]: nit: manages

**Review by paullatzelsperger — APPROVED** [review]
LGTM apart from minor nits.

We could look into introducing another base class `Entity` as superclass of `StatefulEntity`. Analogously we could devise an `InMemoryEntityStore` that in turn is a superclass of `InMemoryEntityStore` and takes those `Entity` objects.
There, we could store things like `Policy`, `ContractDefnition`, `Asset`, etc. That need not be done in this PR though.

**Comment by algattik** [comment]
> LGTM apart from minor nits.
> 
> We could look into introducing another base class `Entity` as superclass of `StatefulEntity`. Analogously we could devise an `InMemoryEntityStore` that in turn is a superclass of `InMemoryEntityStore` and takes those `Entity` objects. There, we could store things like `Policy`, `ContractDefnition`, `Asset`, etc. That need not be done in this PR though.

Yes indeed as I understand that is (at least in part) in scope for @diegogomez-zf in https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/issues/1370.

**Pull request merged**